### PR TITLE
[QNN] Add operator

### PR DESCRIPTION
**Pull request by anijain2305**
Adding QNN Add operator. 

The inputs to QNN Add operator can have different scales and zero points. This PR adds a QNN Add operator that first requantizes the inputs to output scale and shift and then call relay.add. This approach is also used by TF.

### CODE REVIEW

**Comment by u99127** [comment]
This  needs a rebase and squash and some description to go with the pull request.

File: python/tvm/relay/qnn/op/qnn.py [hunk 1]
```diff
@@ -19,14 +19,15 @@
...
+from tvm import relay
 
 def requantize(data,
                input_scale,
                input_zero_point,
                output_scale,
                output_zero_point,
                rounding="TONEAREST",
-               out_dtype="int8"):
+               out_dtype='int8'):
```
Comment (u99127) [inline python/tvm/relay/qnn/op/qnn.py]: Is this really an unrelated change ?

Comment (anijain2305) [reply python/tvm/relay/qnn/op/qnn.py]: Yes, this was unintended change. I will be consistent in using single or double quotes.

**Comment by anijain2305** [comment]
> Is this the right place to finesse the (lhs_scale == rhs_scale and lhs_z_p == rhs_z_p) or is this caught elsewhere ?
> 
> I suspect one could end up with one less requantize step as this is just
> 
> {code}
> output = relay.add (lhs, rhs);
> return requantize (output , ....)
> {code}

Aah, very nice observation. This should be caught here inside the `qnn.add` operator. Let me make the changes.

**Comment by anijain2305** [comment]
@u99127 While working on your comment, I realized that I am not certain the lowering is correct. Give me a couple of days to dig deeper into the TFLite codebase to see what they do. I am not sure if I am handling zero points correctly.

**Comment by u99127** [comment]
> @u99127 While working on your comment, I realized that I am not certain the lowering is correct. Give me a couple of days to dig deeper into the TFLite codebase to see what they do. I am not sure if I am handling zero points correctly.

No worries,  I had another review comment that I missed publishing around testing more cases than just with zero zero_points which is what the tests seem to be doing.

**Comment by anijain2305** [comment]
@u99127 @jackwish @tqchen This is ready for review.

File: python/tvm/relay/qnn/op/qnn.py [hunk 2]
```diff
@@ -72,3 +73,104 @@ def requantize(data,
...
+    requantized_rhs = rhs
+    if rhs_scale != output_scale or rhs_zero_point != output_zero_point:
+        requantized_rhs = requantize(data=rhs,
+                                     input_scale=rhs_scale,
+                                     input_zero_point=rhs_zero_point,
+                                     output_scale=output_scale,
+                                     output_zero_point=output_zero_point,
+                                     out_dtype=in_dtype)
+
+    out = relay.add(requantized_lhs, requantized_rhs)
```
Comment (jackwish) [inline python/tvm/relay/qnn/op/qnn.py]: Similar dtype concern with code above.

File: python/tvm/relay/qnn/op/qnn.py [hunk 3]
```diff
@@ -72,3 +73,104 @@ def requantize(data,
...
+    out = relay.subtract(out, relay.const(output_zero_point, dtype=in_dtype))
```
Comment (jackwish) [inline python/tvm/relay/qnn/op/qnn.py]: Output casting concern.

File: python/tvm/relay/qnn/op/qnn.py [hunk 4]
```diff
@@ -72,3 +73,104 @@ def requantize(data,
...
+    #
+    # RHS looks like a quantized tensor with scale = scale, and zero_point = zp + zp
+    # This can be handled by first subtracting the zero point, followed by requantize with the
+    # output qnn params.
+
+    if lhs_scale == rhs_scale and lhs_zero_point == rhs_zero_point:
+        out = relay.add(lhs, rhs)
+        out = relay.subtract(out, relay.const(lhs_zero_point, dtype=in_dtype))
+        if lhs_scale != output_scale or lhs_zero_point != output_zero_point:
+            out = requantize(data=out,
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply python/tvm/relay/qnn/op/qnn.py]: Yes

**Comment by anijain2305** [comment]
@jackwish Addressed your comments by going to a int32 for addition and then casting back when necessary. Also, added test cases.

File: python/tvm/relay/qnn/op/qnn.py [hunk 5]
```diff
@@ -72,3 +74,120 @@ def requantize(data,
...
+    #
+    #          scale_c * (Q_c - zp_c) = scale * (Q_a - zp) + scale * (Q_b - zp)
+    #          scale_c * (Q_c - zp_c) = scale * (Q_a + Q_b - zp - zp)
+    #
+    # RHS looks like a quantized tensor with scale = scale, and zero_point = zp + zp
+    # This can be handled by first subtracting the zero point, followed by requantize with the
+    # output qnn params. The add op is done in int32 precision.
+
+    if lhs_scale == rhs_scale and lhs_zero_point == rhs_zero_point:
+        lhs = relay.cast(lhs, dtype='int32')
```
Comment (jackwish) [inline python/tvm/relay/qnn/op/qnn.py]: add/sub in int16 should be enough, but not a big deal :)

File: tests/python/relay/test_qnn_add.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    mod = relay.transform.Legalize()(mod)
+    func = mod["main"]
+
+    x_datas = [np.array((140, 153, 165, 178)).reshape((1,4)),
+               np.array((25, 153, 178, 216)).reshape((1,4)),
+               np.array((25, 153, 216, 165)).reshape((1,4))]
+    y_datas = [np.array((204, 178, 165, 140)).reshape((1,4)),
+               np.array((204, 178, 191, 25)).reshape((1,4)),
+               np.array((204, 178, 25, 191)).reshape((1,4))]
+    golden_outputs = [np.array((217,204,203,191)).reshape((1, 4)),
```
Comment (jackwish) [inline tests/python/relay/test_qnn_add.py]: A bit curious, are these data coming from TFLite computing results, or manually computed?

Comment (anijain2305) [reply tests/python/relay/test_qnn_add.py]: TFLite adds quantize and dequantize. I am trapping the numbers after they have been quantized and before they are getting dequantized. So, these are from TFLite but not the actual GTest that you see, but the internals of it.

Comment (jackwish) [reply tests/python/relay/test_qnn_add.py]: Yes, thank you for the explanation.

File: python/tvm/relay/qnn/op/qnn.py (hunk 5 above)
Comment (anijain2305) [reply python/tvm/relay/qnn/op/qnn.py]: Tried that :) Currently, it fails before requantize input can only be (uint8, int8, int32). I think for now, it should be ok. If we see more demand of int16, we can add support across all the QNN ops.

Comment (jackwish) [reply python/tvm/relay/qnn/op/qnn.py]: Yes, that is far from a blocking issue :) Thank you.

**Comment by Leo-arm** [comment]
It seems that the normalisation of the quantisation parameters would be the same for a number of different operators. If that is the case, does it make sense to factor this out and maybe put this in a pass? That would avoid having to implement this for each operator.
Do we know what the quantization parameters are at each point in the Relay graph?

**Comment by anijain2305** [comment]
> It seems that the normalisation of the quantisation parameters would be the same for a number of different operators. If that is the case, does it make sense to factor this out and maybe put this in a pass? That would avoid having to implement this for each operator.
> Do we know what the quantization parameters are at each point in the Relay graph?

@Leo-arm this is a very good point. For background, there are 2 parallel efforts in TVM community right now

* Automatic Quantization - Quantization in Relay. FP32 model comes in and Relay pass quantizes it.
* Supporting Pre-quantized models - The quantization happens outside of Relay and TVM. (This PR belongs to this category). Here, the main goal is to create new QNN ops that will be converted to existing Relay ops.

We can share the HW schedules between these two options.

What you suggested is almost what happens today in the Automatic quantization project.

For doing this in pre-quantized models, it is somewhat tricky because it happens on an op-by-op basis (`requantize` happens before in `concatenate`, but it happens later in `conv2d`. The zero point handling is different for `add` vs `multiply`). But, there are certainly some patterns that can be exploited to make it easier. For example, we can have a helper function that requantizes the inputs and then calls the main op, and all the QNN ops that follow this pattern can rely on the helper function. For now, this seems like a good intermediate point. As we get more ops and pattens, we can refactor those using helper functions.

**Comment by u99127** [comment]
Assume below that ip0, ip1 and op are all 8 bit tensors with identical zero points and all the other cases. 

{code}
op = add (ip0, ip1);
{code}

now gets lowered into : 

{code}
tmp1 = recast (ip0, 32 bits)
tmp2 = recast (ip1, 32 bits)
tmp3 = add (tmp1, tmp2)
op = requantize (tmp3)
{code}

Am I right in assuming that the tflite parser directly lowers to this level ? Is there any reason why the alternate option of having 8 bit tensor operations in relay has been ignored ?

**Comment by anijain2305** [comment]
> Am I right in assuming that the tflite parser directly lowers to this level ? Is there any reason why the alternate option of having 8 bit tensor operations in relay has been ignored?

* If the original op did not have any quantization params, if It was just plain int8 op, we will directly call `relay.add` instead of `qnn.add`.
* In case of original op with quantization params, I would still like to cast. Since in quantization, the integers are just a proxy for FP32 numbers, overflow (with just `int8`) can lead to a bad proxy FP32 number. So, I have opted for a saturating behavior (manifested by `requantize`).

**Comment by anijain2305** [comment]
@u99127 Please review again to ensure that your questions are answered :)

@vinx13 Ping for review.

**Comment by anijain2305** [comment]
Do not submit yet. Will move the codebase to C++ to avoid calling InferType.

**Comment by anijain2305** [comment]
Moved to C++. Removed WIP tag

**Comment by anijain2305** [comment]
@jackwish Can you please review? I have rebased to master.

**Comment by anijain2305** [comment]
@zhiics @vinx13 Can you please review when you get time.

File: src/relay/qnn/op/add.cc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+  // Get the input dtype and shape.
+  CHECK_EQ(arg_types.size(), 3);
+  auto tensor_type = arg_types[0].as<TensorTypeNode>();
+  auto input_dtype = tensor_type->dtype;
+  auto input_shape = tensor_type->shape;
+
+  // FIXME (anijain2305) - The lowering can be further optimized. Instead of inserting requantize in
+  // the start, we can insert requantize at the end if and only if all the input tensors have same
+  // qnn params. This can be done in future.
```
Comment (jackwish) [inline src/relay/qnn/op/add.cc]: I guess that same `scale` can lead to requantize after ADD, `zero point` can be safely subtracted :)

File: src/relay/qnn/op/add.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+  // Requantize RHS if necessary.
+  auto requantized_rhs = rhs;
+  if (rhs_scale != output_scale || rhs_zero_point != output_zero_point) {
+    requantized_rhs = Requantize(rhs, input_shape, rhs_scale, rhs_zero_point, output_scale,
+                                 output_zero_point, input_dtype);
+  }
+
+  // Upcast to maintain precision.
+  requantized_lhs = Cast(requantized_lhs, Int(32));
```
Comment (jackwish) [inline src/relay/qnn/op/add.cc]: It seems that result of two int8 subtracting can be hold in int16? But not big deal :)

File: src/relay/qnn/op/add.cc (hunk 7 above)
Comment (anijain2305) [reply src/relay/qnn/op/add.cc]: Let me change the comment.

File: src/relay/qnn/op/add.cc (hunk 8 above)
Comment (anijain2305) [reply src/relay/qnn/op/add.cc]: Yes, currently Requantize does not support Int16. So, we can skip it for now. If we see int16 need later on, we can start supporting it across all ops.

**Comment by zhiics** [comment]
Thanks @anijain2305 @jackwish @u99127, this is now merged.
